src/test/java/htsjdk/samtools/util/IOUtilTest.java

```diff
@@ -520,4 +527,240 @@ public void testIsBlockCompressedOnJimfs(Path file, boolean checkExtension, bool
+        final long origSize = Files.size(file);
+        long previousSize = origSize;
+        for (int compressionLevel = 1; compressionLevel <= 9; compressionLevel++) {
+            final Path outFile = Files.createTempFile("tmp", extension);
+            outFile.toFile().deleteOnExit();
+            IOUtil.setCompressionLevel(compressionLevel);
+            Assert.assertEquals(IOUtil.getCompressionLevel(), compressionLevel);
+            final InputStream inStream = IOUtil.openFileForReading(file);
+
+            try (final OutputStream outStream = IOUtil.openFileForWriting(outFile.toFile())) {
+
+
+                IOUtil.transferByStream(inStream, outStream, origSize);
+            }
+
+
+            final long newSize = Files.size(outFile);
+            if (compressionLevel <= lastDifference) {
+                Assert.assertTrue(previousSize > newSize);
+            } else {
+                Assert.assertTrue(previousSize >= newSize);
+            }
+            previousSize = newSize;
+
+        }
+
+    }
+
+    @DataProvider
+    public static Object[][] badCompressionLevels() {
+        return new Object[][]{
+                {-1},
+                {10}
+        };
+
+    }
+
+    @Test(dataProvider = "badCompressionLevels", expectedExceptions = {IllegalArgumentException.class})
+    public void testCompressionLevelExceptions(final int compressionLevel) {
+        IOUtil.setCompressionLevel(compressionLevel);
+    }
+
+    @DataProvider
+    public static Object[][] filesToCopy() {
+        return new Object[][]{
+                {TEST_VARIANT_DIR.resolve("test1.vcf")},
+                {TEST_DATA_DIR.resolve("ipsum.txt")}
+
+        };
+    }
+
+    @Test(dataProvider = "filesToCopy")
+    public void testCopyFile(final Path file) throws IOException {
+        final Path outFile = Files.createTempFile("tmp", ".tmp");
+        outFile.toFile().deleteOnExit();
+        IOUtil.copyFile(file.toFile(), outFile.toFile());
+
+
+        final Stream<String> stream = Files.lines(file);
+        final Stream<String> outStream = Files.lines(outFile);
+
+        final Iterator<String> itStream = stream.iterator();
+        final Iterator<String> itOutStream = outStream.iterator();
+
+        while (itStream.hasNext() && itOutStream.hasNext()) {
+            Assert.assertEquals(itStream.next(), itOutStream.next());
+        }
+        Assert.assertFalse(itStream.hasNext());
+        Assert.assertFalse(itOutStream.hasNext());
+    }
+
+    @Test(dataProvider = "filesToCopy", expectedExceptions = {SAMException.class})
+    public void testCopyFileReadException(final Path file) throws IOException {
+        final Path outFile = Files.createTempFile("tmp", ".tmp");
+        outFile.toFile().deleteOnExit();
+        file.toFile().setReadable(false);
+        try {
+            IOUtil.copyFile(file.toFile(), outFile.toFile());
+        } catch (SAMException e) {
+            file.toFile().setReadable(true);

```

Comment:
why change the readability status here?

Reply:
if this is needed then put a comment.

Reply:
you want a finally here not a catch.

Reply:
yup, you're right should be a finally.  For reference, need to change back to readable so that other tests which need to read this file later will not fail (along with just being best to leave things as they were before the test).  adding explanation as comment in code.